src/batch/src/executor/join/lookup_join.rs

```diff
@@ -0,0 +1,576 @@
+            column_ids: (0..self.table_desc.columns.len())
+                .into_iter()
+                .map(|x| x as i32)
+                .collect(), // Scan all the columns
+            scan_range,
+            vnode_bitmap: None,
+        }))
+    }
+
+    /// Creates all the `ProstExchangeSource` that will be sent to the `ExchangeExecutor` using
+    /// the source templates.
+    fn build_prost_exchange_sources(&self, cur_row: &RowRef) -> Result<Vec<ProstExchangeSource>> {
+        let Plan(inner_template_plan) = self.source_templates[0].get_local_execute_plan().unwrap();
+
+        let local_execute_plan = LocalExecutePlan {
+            plan: Some(PlanFragment {
+                root: Some(PlanNode {
+                    children: vec![],
+                    identity: Uuid::new_v4().to_string(),
+                    node_body: Some(self.create_row_seq_scan_node(cur_row)?),
+                }),
+                exchange_info: None,
+            }),
+            epoch: inner_template_plan.epoch,
+        };
+
+        let prost_exchange_sources = self
+            .source_templates
+            .iter()
+            .map(|source| ProstExchangeSource {
+                task_output_id: Some(source.get_task_output_id().unwrap().clone()),
+                host: Some(source.get_host().unwrap().clone()),
+                local_execute_plan: Some(Plan(local_execute_plan.clone())),
+            })
+            .collect();
+
+        Ok(prost_exchange_sources)
+    }
+}
+
+#[async_trait::async_trait]
+impl<C: BatchTaskContext> ProbeSideSourceBuilder for ProbeSideSource<C> {
+    /// Builds and returns the `ExchangeExecutor` used for the probe side of the
+    /// `LookupJoinExecutor`.
+    async fn build_source(&self, cur_row: &RowRef) -> Result<BoxedExecutor> {
+        let sources = self.build_prost_exchange_sources(cur_row)?;
+
+        let exchange_node = NodeBody::Exchange(ExchangeNode {
+            sources,
+            input_schema: self.probe_side_schema.to_prost(),
+        });
+
+        let plan_node = PlanNode {
+            children: vec![],
+            identity: "LookupJoinExchangeExecutor".to_string(),
+            node_body: Some(exchange_node),
+        };
+
+        let task_id = self.task_id.clone();
+
+        let executor_builder =
+            ExecutorBuilder::new(&plan_node, &task_id, self.context.clone(), self.epoch);
+
+        executor_builder.build().await
+    }
+}
+
+/// Lookup Join Executor.
+/// High-level idea:
+/// 1) Iterate through each row in the build side
+/// 2) For each row R, find the rows that match with R on the join conditions on the probe side by
+/// using the `GatherExecutor` to query the probe side
+/// 3) Join R and the matching rows on the probe side together with either inner or left outer join
+/// 4) Repeat 2-3) for every row in the build side
+pub struct LookupJoinExecutor<P> {
+    join_type: JoinType,
+    condition: Option<BoxedExpression>,
+    build_child: Option<BoxedExecutor>,
```

Comment:
Why is it an `Option`?

Reply:
I'm not sure if it needs to be an `Option`, I was running into some partial moving errors that were fixed when I placed the build_child's `BoxedExecutor`. into an `Option`. This is also what `HashJoinExecutor` does, too.